File: golang/ch1-ArrayStrings/answers/str_compress/str_compress.go
/*
Implement a method to perform basic string compress using the counts of repeated characters. If the compressed string would not become smaller than the original string, your method should return the original string.

You can assume str has only uppercase and lowercase letters.

Example:
    Input: "aabcccccaaa"
    Output: "a2b1c5a3"

Hints: 92, 110
*/
package str_compress

import "fmt"

func StrCompress(s string) string {

	var cnt int8
	var retstr string
	if s == "" {
		return s
	}
	last := rune(s[0])

	for index, curr := range s {
		if curr == last && index != len(s)-1 {
			cnt += 1
		} else {
			if index == len(s)-1 {
				if curr == last {
					cnt += 1
				} else {
					last = curr
					cnt = 1
				}
			}
			retstr = fmt.Sprintf("%v%c%v", retstr, last, cnt)
			last = curr
			cnt = 1
		}
	}

	if len(s) <= len(retstr) {
		return s
	}
	return retstr

}

File: ch1-ArraysStrings/questions/str_compress.py
'''
Implement a method to perform basic string compress using the counts of repeated characters. If the compressed string would not become smaller than the original string, your method should return the original string.

You can assume str has only uppercase and lowercase letters.

Example:
    Input: "aabcccccaaa"
    Output: "a2b1c5a3"

Hints: 92, 110
'''


if __name__ == '__main__':
    assert str_compress2('a') == 'a'
    assert str_compress2('aaa') == 'a3'
    assert str_compress2('aabbccc') == 'a2b2c3'
    assert str_compress2('blahblahblah') == 'blahblahblah'
    assert str_compress2('mmmmmhhhhhccccckkkkk') == 'm5h5c5k5'
    assert str_compress2('/////') == '/5'
    assert str_compress2('     ') == ' 5'
    assert str_compress2('aabcccccaaa') == 'a2b1c5a3'
